pkg/components/internal/testutil/jsonpath.go

```diff
@@ -0,0 +1,91 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testutil
+
+import (
+	"reflect"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	yamlserializer "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/client-go/util/jsonpath"
+)
+
+// getUnstructredObj inputs any Kubernetes manifest in YAML format and returns an object of type
+// `unstructured.Unstructured`. This object has many methods that can be used by the consumer to
+// extract metadata of the Kubernetes manifest.
+func getUnstructredObj(t *testing.T, yamlObj string) *unstructured.Unstructured {
+	u := &unstructured.Unstructured{}
+
+	// Decode YAML into `unstructured.Unstructured`.
+	dec := yamlserializer.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	if _, _, err := dec.Decode([]byte(yamlObj), nil, u); err != nil {
+		t.Fatalf("Converting config to unstructured.Unstructured: %v", err)
+	}
+
+	return u
+}
+
+// getValFromObject takes JSON path as string and an object of type unstructured.Unstructured. This
+// function returns object of type reflect.Value at that JSON path.
+func getValFromObject(t *testing.T, jp string, obj *unstructured.Unstructured) reflect.Value {
+	jPath := jsonpath.New("parse")
+	if err := jPath.Parse(jp); err != nil {
+		t.Fatalf("Parsing JSONPath: %v", err)
+	}
+
+	v, err := jPath.FindResults(obj.Object)
+	if err != nil {
+		t.Fatalf("Finding results using JSONPath in the YAML file: %v", err)
+	}
+
+	if len(v) == 0 || len(v[0]) == 0 {
+		t.Fatalf("No result found")
+	}
+
+	return v[0][0]
+}
+
+// getJSONPathValue extracts an object at a JSON path from a YAML config, and returns interface
+// object.
+func getJSONPathValue(t *testing.T, yamlConfig string, jsonPath string) interface{} {
+	u := getUnstructredObj(t, yamlConfig)
+	got := getValFromObject(t, jsonPath, u)
+
+	switch got.Kind() { //nolint:exhaustive
+	case reflect.Interface:
+		// TODO: Add type switch here for concrete types.
+		return got.Interface()
+	default:
+		t.Fatalf("Unknown type of the object extracted from the converted YAML: %v", got.Kind())
+	}
+
+	return nil
+}
+
+// MatchJSONPathStringValue is a helper function for component's unit tests. It compares a string
+// value of a JSON path in a YAML config.
+func MatchJSONPathStringValue(t *testing.T, yamlConfig string, jsonPath string, expected string) {
+	obj := getJSONPathValue(t, yamlConfig, jsonPath)
+
+	got, ok := obj.(string)
+	if !ok {
+		t.Fatalf("Value is not string: %#v", obj)
+	}
+
+	if got != expected {
+		t.Fatalf("Expected: %s, Got: %s", expected, got)
+	}
```

Comment:
I am strongly against extracting this type of checks to a helper function. In my opinion these checks are **the** thing we want to include in each individual unit test, even if it makes things a bit repetitive. The reader should be able to figure out what a unit tests does without jumping around the codebase.

Reply:
Right now it looks simplistic because we are just comparing strings. We would also add support for strings, numbers, arrays of concrete types then it will become complicated. So I would like to keep it here. So it becomes easy to extend.

Reply:
IMO we shouldn't refactor things *now* because things *may* change in the future. We will be a lot smarter regarding how to refactor and simplify things when we actually see the problem. If the reader has to jump to a different source file when reading a unit test to understand what the test does, I think it's not ideal.